* Data
** +++ TCGA clinical attributes, gene expression, CNA segments
** +++ MC3 variants - https://www.synapse.org/#!Synapse:syn7214402/files/
** ++ G2P (talk to walsh)
** + ML signatures
** + Literome - https://ohsu.app.box.com/files/0/f/21493766712/G2P_Data
** pubmed - ftp://ftp.ncbi.nlm.nih.gov/pubmed/baseline
** PDB
** support DMS queries (georgia)
** ETL of existing data (TCGA, signatures): work with DMS
** Import SIF (pathways)
* Machine Learning
** + Create schema for SVM
** merge all ML libs in bmeg org into ladder
* Visualization
** + all state of application encoded in window location (redux and router)
** + onco.js add sequence for edges
** fix response curves on cohort page to account for inconsistent x-axis points
** add more visualizations
*** cohort page: show what are the top 10 mutated genes in this cohort - bar chart
*** precalculate counts for each type, if there are less than 30 show drop down
* Protograph
** ++ Create support for ingesting messages into edges and edge properties
** combine protograph with protobuf schemas to fill out properties and types for describing the full graph schema
** create conformance test
* Ophion
** ++ support indexing
** ++ sync to Kyle's protobuf - add edge label step and write api to ophion
** pass and extend Ophion conformance test
* Server
** +++ add titan connection pool
** +++ stream results as you get them
** +++ add elastic search for vertex types (case insensitve with prefix matching)
** improve caching in server for queries (proxy to memcache for round-robin expiration)
* Agents
** + setup agent interface
** + run Heskett's clustering agent
** make agent for adding edges between samples and genes
** align signature code with protograph, make an agent
** create tests for protograph
* Site
** institute a site update page with one card per update
** give some attention to the other tabs
* Support
** Write docs for everyone! and everything!
** help Chris
