* Data
** ++ TCGA clinical attributes, gene expression, CNA segments
** ++ MC3 variants - https://www.synapse.org/#!Synapse:syn7214402/files/
** + ML signatures
** pubmed - ftp://ftp.ncbi.nlm.nih.gov/pubmed/baseline
** Literome - https://ohsu.app.box.com/files/0/f/21493766712/G2P_Data
** PDB
** G2P (talk to walsh)
** support DMS queries (georgia)
** fill in kafka topics with bioschema messages
** ETL of existing data (TCGA, signatures): work with DMS
** Make edges between samples and mutations (mutation summary)
** Import SIF (pathways)
* Machine Learning
** + Create schema for SVM
** merge all ML libs in bmeg org into ladder
* Visualization
** +++ add response curves to cohort page
** + all state of application encoded in window location (redux and router)
** + onco.js add sequence for edges
** add more visualizations
*** cohort page: show what are the top 10 mutated genes in this cohort - bar chart
*** precalculate counts for each type, if there are less than 30 show drop down
* Protograph
** ++ Create support for ingesting messages into edges and edge properties
** combine protograph with protobuf schemas to fill out properties and types for describing the full graph schema
** create conformance test
* Ophion
** + support indexing
** + sync to Kyle's protobuf - add edge label step and write api to ophion
** pass and extend Ophion conformance test
* Server
** +++ add titan connection pool
** +++ add elastic search for vertex types (case insensitve with prefix matching)
** + stream results as you get them
** add some kind of caching in server for queries
* Agents
** + setup agent interface
** + run Heskett's clustering agent
** make agent for adding edges between samples and genes
** align signature code with protograph, make an agent
** create tests for protograph
* Support
** Write docs for everyone! and everything!
** help Chris
